### PR TITLE
[FEATURE] Add isRemoveDisplayNoneEnabled option + tests

### PR DESCRIPTION
**Pull request by timothyasp**
This adds an options parameter and some tests for a `isRemoveDisplayNoneEnabled` option
which specifies if the Emogrifier parser will remove elements with the style
"display: none".

Unfortunatly, the nice little optimization has some consiquences.

When writing responsive email templates, "display: none" is used often to hide
certain sections of the template for desktop email clients and show them in
phone clients.

I can see how it'd be nice to clean up elements that will not be displayed in
some cases, so that's why I made it an options field.

### CODE REVIEW

File: Classes/Emogrifier.php [hunk 1]
```diff
@@ -122,6 +122,12 @@ class Emogrifier
     private $isStyleBlocksParsingEnabled = true;
 
     /**
+     * Determines whether elements with the `display: none` property are
+     * removed from the DOM.
+     */
+    private $isRemoveDisplayNoneEnabled = true;
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Actually, the option name is a bit misleading. As it is about removing/keeping nodes that are display: none (and not about removing/keeping the display: none itself), I propose the name "shouldKeepInvisibleNodes". What do you think?

File: Classes/Emogrifier.php [hunk 2]
```diff
@@ -122,6 +122,12 @@ class Emogrifier
     private $isStyleBlocksParsingEnabled = true;
 
     /**
+     * Determines whether elements with the `display: none` property are
+     * removed from the DOM.
+     */
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please add the missing @var bool.

File: Classes/Emogrifier.php [hunk 3]
```diff
@@ -294,16 +300,18 @@ public function emogrify()
...
-        );
-        // The checks on parentNode and is_callable below ensure that if we've deleted the parent node,
-        // we don't try to call removeChild on a nonexistent child node
-        if ($nodesWithStyleDisplayNone->length > 0) {
-            /** @var \DOMNode $node */
-            foreach ($nodesWithStyleDisplayNone as $node) {
-                if ($node->parentNode && is_callable(array($node->parentNode,'removeChild'))) {
-                    $node->parentNode->removeChild($node);
+        if ($this->isRemoveDisplayNoneEnabled) {
+            $nodesWithStyleDisplayNone = $xpath->query(
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please let's move the code within the IF to a separate, well-named private method.

File: Classes/Emogrifier.php [hunk 4]
```diff
@@ -294,16 +300,18 @@ public function emogrify()
...
+                '//*[contains(translate(translate(@style," ",""),"NOE","noe"),"display:none")]'
+            );
+            // The checks on parentNode and is_callable below ensure that if we've deleted the parent node,
+            // we don't try to call removeChild on a nonexistent child node
+            if ($nodesWithStyleDisplayNone->length > 0) {
+                /** @var $node \DOMNode */
+                foreach ($nodesWithStyleDisplayNone as $node) {
+                    if ($node->parentNode && is_callable(array($node->parentNode,'removeChild'))) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please add the missing space after the comma.

File: Tests/Unit/EmogrifierTest.php [hunk 5]
```diff
@@ -1090,4 +1090,45 @@ public function emogrifyMergesCssWithMixedUnits()
             $this->subject->emogrify()
         );
     }
+
+    /**
+     * @test
+     */
+    public function emogrifyRemovesDisplayNoneElements()
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: Please add the precondition for this behavior to the test name (also for the other test): emogrifyFor...RemovesInvisibleElement

File: Tests/Unit/EmogrifierTest.php [hunk 6]
```diff
@@ -1090,4 +1090,45 @@ public function emogrifyMergesCssWithMixedUnits()
...
+    {
+        $css = 'div.foo { display: none; }';
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: We also should have tests for both cases: When the display: none comes from a style node, and when it comes from a style attribute of the element.

File: Classes/Emogrifier.php (hunk 1 above)
Comment (timothyasp) [inline Classes/Emogrifier.php]: Sure, that sounds fine.

**Comment by timothyasp** [comment]
@oliverklee I think i've made all the requested changes.  Thanks for the feedback!

File: Classes/Emogrifier.php [hunk 7]
```diff
@@ -338,6 +331,16 @@ public function disableStyleBlocksParsing()
     }
 
     /**
+     * Disables the removal of elements with `display: none` properties
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please add the period at the end of the sentence.

File: Classes/Emogrifier.php [hunk 8]
```diff
@@ -418,6 +421,34 @@ public function removeUnprocessableHtmlTag($tagName)
...
+      */
+    private function removeInvisibleNodes(\DOMXPath $xpath)
+    {
+        $nodesWithStyleDisplayNone = $xpath->query(
+            '//*[contains(translate(translate(@style," ",""),"NOE","noe"),"display:none")]'
+        );
+        // The checks on parentNode and is_callable below ensure that if we've deleted the parent node,
+        // we don't try to call removeChild on a nonexistent child node
+        if ($nodesWithStyleDisplayNone->length > 0) {
+            /** @var $node \DOMNode */
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please let's change the order of the elements within this @var comment: First the type, then the name. This will make the comment consistent with the other @var comments for local variables.

File: Classes/Emogrifier.php [hunk 9]
```diff
@@ -418,6 +421,34 @@ public function removeUnprocessableHtmlTag($tagName)
...
+      * @return void
+      */
+    private function removeInvisibleNodes(\DOMXPath $xpath)
+    {
+        $nodesWithStyleDisplayNone = $xpath->query(
+            '//*[contains(translate(translate(@style," ",""),"NOE","noe"),"display:none")]'
+        );
+        // The checks on parentNode and is_callable below ensure that if we've deleted the parent node,
+        // we don't try to call removeChild on a nonexistent child node
+        if ($nodesWithStyleDisplayNone->length > 0) {
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Now that this is within a separate nice method, we can change this to a guard clause:

if ... === 0 {
  return;
}

This will allow us to drop one level of indentation for the following code. Could you please do this?

File: Tests/Unit/EmogrifierTest.php [hunk 10]
```diff
@@ -1090,4 +1090,46 @@ public function emogrifyMergesCssWithMixedUnits()
...
+        $this->assertContains(
+            $expected,
+            $this->subject->emogrify()
+        );
+    }
+
+    /**
+     * @test
+     */
+    public function emogrifyPreservesInvisibleElements()
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: Please add the precondition to the test name:

emogrifyAfterDisableInvisibleNodeRemovalPreservesInvisibleElements

File: Tests/Unit/EmogrifierTest.php [hunk 11]
```diff
@@ -1090,4 +1090,46 @@ public function emogrifyMergesCssWithMixedUnits()
...
     }
+
+    /**
+     * @test
+     */
+    public function emogrifyRemovesInvisibleElements()
+    {
+        $css = 'div.foo { display: none; }';
+        $html = $this->html5DocumentType . self::LF .
+            '<html><body><div class="bar"></div><div class="foobar" style="display: none;"></div>'
```
Comment (oliverklee) [inline Tests/Unit/EmogrifierTest.php]: Actually, these should be two separate tests: one for the case that the display: none comes from the CSS via setCss, and one for the case that the display: none comes from the style attribute. I propose the following two test method names:

emogrifyByDefaultRemovesElementsWithDisplayNoneInStyleAttribute
emogrifyByDefaultRemovesElementsWithDisplayNoneFromExternalCss

What do you think?

**Comment by oliverklee** [comment]
Could you also please write a short documentation for this new feature for the README?

**Comment by oliverklee** [comment]
ping

**Comment by timothyasp** [comment]
Made the requested changes.  The test failure looks Github related?  Travis couldn't connect to github.  I'm not sure how to requeue the test - @oliverklee could you do that?

**Comment by oliverklee** [comment]
I've run the tests on this branch locally (as long as Travis does not play along). On your branch, I get the following failure:

    klee@gonzales:~/src/php/emogrifier$ vendor/bin/phpunit Tests/
    PHPUnit 4.6.7 by Sebastian Bergmann and contributors.
    
    ...............................................................  63 / 155 ( 40%)
    ............................................................... 126 / 155 ( 81%)
    ...........................F.
    
    Time: 74 ms, Memory: 5.00Mb
    
    There was 1 failure:
    
    1) Pelago\Tests\Unit\EmogrifierTest::emogrifyByDefaultRemovesElementsWithDisplayNoneInStyleAttribute
    Failed asserting that '<!DOCTYPE html>
    <html><body>
    <div class="bar"></div>
    <div class="foo"></div>
    </body></html>
    ' contains "<div class="bar"></div><div class="foo"></div>".
    
    /home/klee/src/php/emogrifier/Tests/Unit/EmogrifierTest.php:1130
    
    FAILURES!
    Tests: 155, Assertions: 155, Failures: 1.

Do the tests run fine on your local machine?

**Comment by oliverklee** [comment]
And could you please prefix the commit message subject line with [FEATURE]?

File: Classes/Emogrifier.php [hunk 12]
```diff
@@ -418,6 +421,37 @@ public function removeUnprocessableHtmlTag($tagName)
...
+      * @param \DOMXPath $xpath
+      *
+      * @return void
+      */
+    private function removeInvisibleNodes(\DOMXPath $xpath)
+    {
+        $nodesWithStyleDisplayNone = $xpath->query(
+            '//*[contains(translate(translate(@style," ",""),"NOE","noe"),"display:none")]'
+        );
+
```
Comment (oliverklee) [inline Classes/Emogrifier.php]: Please remove this blank line so that the guard clause is a single visible block.

**Comment by timothyasp** [comment]
I address all your concerns - still looks like Travis isn't working with github.

**Comment by oliverklee** [comment]
I've tested the PR locally, fixed the unit tests (the DIVs were not nested, which made the assertContains not find any potential problems reliably), merged locally, and pushed.

Thanks for contributing, @timothyasp! :heart:
